module/zfs/brt.c

```diff
@@ -0,0 +1,1886 @@
+    uint64_t dsize)
+{
+	uint64_t idx;
+
+	ASSERT(RW_WRITE_HELD(&brt->brt_lock));
+	ASSERT(brtvd != NULL);
+	ASSERT(brtvd->bv_entcount != NULL);
+
+	brt->brt_savedspace -= dsize;
+	brtvd->bv_savedspace -= dsize;
+	brtvd->bv_meta_dirty = TRUE;
+
+	if (bre->bre_refcount > 0) {
+		return;
+	}
+
+	brt->brt_usedspace -= dsize;
+	brtvd->bv_usedspace -= dsize;
+
+	idx = bre->bre_offset / brt->brt_rangesize;
+	ASSERT3U(idx, <, brtvd->bv_size);
+
+	ASSERT(brtvd->bv_totalcount > 0);
+	brtvd->bv_totalcount--;
+	brt_vdev_entcount_dec(brtvd, idx);
+	brtvd->bv_entcount_dirty = TRUE;
+	idx = idx / BRT_BLOCKSIZE / 8;
+	BT_SET(brtvd->bv_bitmap, idx);
+
+#ifdef ZFS_DEBUG
+	brt_vdev_dump(brt);
+#endif
+}
+
+static void
+brt_vdev_sync(brt_t *brt, brt_vdev_t *brtvd, dmu_tx_t *tx)
+{
+	dmu_buf_t *db;
+	brt_vdev_phys_t *bvphys;
+
+	ASSERT(brtvd->bv_meta_dirty);
+	ASSERT(brtvd->bv_mos_brtvdev != 0);
+	ASSERT(dmu_tx_is_syncing(tx));
+
+	VERIFY0(dmu_bonus_hold(brt->brt_mos, brtvd->bv_mos_brtvdev, FTAG, &db));
+
+	if (brtvd->bv_entcount_dirty) {
+		/*
+		 * TODO: Walk brtvd->bv_bitmap and write only the dirty blocks.
+		 */
+		dmu_write(brt->brt_mos, brtvd->bv_mos_brtvdev, 0,
+		    brtvd->bv_size * sizeof (brtvd->bv_entcount[0]),
+		    brtvd->bv_entcount, tx);
+		memset(brtvd->bv_bitmap, 0, BT_SIZEOFMAP(brtvd->bv_nblocks));
+		brtvd->bv_entcount_dirty = FALSE;
+	}
+
+	dmu_buf_will_dirty(db, tx);
+	bvphys = db->db_data;
+	bvphys->bvp_mos_entries = brtvd->bv_mos_entries;
+	bvphys->bvp_size = brtvd->bv_size;
+	if (brtvd->bv_need_byteswap) {
+		bvphys->bvp_byteorder = BRT_NON_NATIVE_BYTEORDER;
+	} else {
+		bvphys->bvp_byteorder = BRT_NATIVE_BYTEORDER;
+	}
+	bvphys->bvp_totalcount = brtvd->bv_totalcount;
+	bvphys->bvp_rangesize = brt->brt_rangesize;
+	bvphys->bvp_usedspace = brtvd->bv_usedspace;
+	bvphys->bvp_savedspace = brtvd->bv_savedspace;
+	dmu_buf_rele(db, FTAG);
+
+	brtvd->bv_meta_dirty = FALSE;
+}
+
+static void
+brt_vdevs_alloc(brt_t *brt, boolean_t load)
+{
+	brt_vdev_t *brtvd;
+	uint64_t vdevid;
+
+	brt_wlock(brt);
+
+	brt_vdevs_expand(brt, brt->brt_spa->spa_root_vdev->vdev_children);
+
+	if (load) {
+		for (vdevid = 0; vdevid < brt->brt_nvdevs; vdevid++) {
+			brtvd = &brt->brt_vdevs[vdevid];
+			ASSERT(brtvd->bv_entcount == NULL);
+
+			brt_vdev_load(brt, brtvd);
+		}
+	}
+
+	if (brt->brt_rangesize == 0) {
+		brt->brt_rangesize = BRT_RANGESIZE;
+	}
+
+	brt_unlock(brt);
+}
+
+static void
+brt_vdevs_free(brt_t *brt)
+{
+	brt_vdev_t *brtvd;
+	uint64_t vdevid;
+
+	brt_wlock(brt);
+
+	for (vdevid = 0; vdevid < brt->brt_nvdevs; vdevid++) {
+		brtvd = &brt->brt_vdevs[vdevid];
+		if (brtvd->bv_initiated)
+			brt_vdev_dealloc(brt, brtvd);
+	}
+	kmem_free(brt->brt_vdevs, sizeof (brt_vdev_t) * brt->brt_nvdevs);
+
+	brt_unlock(brt);
+}
+
+static void
+brt_entry_fill(const blkptr_t *bp, brt_entry_t *bre, uint64_t *vdevidp)
+{
+
+	memset(bre, 0, sizeof (*bre));
+	bre->bre_offset = DVA_GET_OFFSET(&bp->blk_dva[0]);
```

Comment:
May be cosmetic or not, but I am not a big fan of excessive initialization.  I'd replace memset() here with explicit `bre->bre_refcount = 0;`, since some paths like brt_entry_alloc() seem to read it after this.  And in brt_entry_alloc() I'd just remove memset().  I don't see a point to wipe often not used 24 bytes of bre_node.  There is no equivalent of list_link_active() in AVL IIRC.

Reply:
I did proposed changes, but personally I don't mind this extra initialization. IMHO it makes the code safer to maintain, ie. if we decide to add more fields to this structure and forget to update those initialization we have a higher chance to end up with a buggy code if we initialize the fields individually.

Reply:
Kind of yes, but when you add field, the first thing you'd think about reasonable initialization.  For large and rarely used structures memset() may be cleaner that zillion of assignments, but this small function may be called several times per block in places that are expected to be fast.